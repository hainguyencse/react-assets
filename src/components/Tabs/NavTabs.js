import React, { Component } from 'react';
import { getTabId } from '.';

const Tab = ({ activeTab, onTabChange, label, id }) => (
  <li className={activeTab === id ? 'active' : ''}>
    <a style={{ cursor: 'pointer' }} onClick={() => onTabChange(id)}>{label}</a>
  </li>
);

class NavTabs extends Component {
  getTabId = (label, idx) => {
    const { tabIds } = this.props;

    if (tabIds && tabIds[idx]) {
      return tabIds[idx];
    }
    return getTabId(label);
  };

  render() {
    const { tabLabels, onTabChange, activeTab } = this.props;

    return (
      <ul className="nav nav-tabs">
        {tabLabels.map((label, idx) => {
          const tabId = this.getTabId(label, idx);
          return (
            <Tab
              key={tabId}
              activeTab={activeTab}
              onTabChange={onTabChange}
              label={label}
              id={tabId}
            />
          );
        })}
      </ul>
    );
  }
}

export default NavTabs;
