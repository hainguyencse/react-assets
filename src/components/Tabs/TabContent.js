import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { TabContext } from '.';

const TabContent = ({ tabId, component: Component, ...rest }) => (
  <TabContext.Consumer>
    {activeTab => (
      <div className={cx(['tab-pane', activeTab === tabId && 'active'])}>
        <Component {...rest} />
      </div>
    )}
  </TabContext.Consumer>
);

TabContent.propTypes = {
  /**
   * Id of the tab content
   * Must exist in the list of ids provided to tab
   */
  tabId: PropTypes.string.isRequired,
  /**
   * The component to render inside
   */
  component: PropTypes.func.isRequired,
};

export default TabContent;
