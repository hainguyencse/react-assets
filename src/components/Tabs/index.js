import React from 'react';
import PropTypes from 'prop-types';
import NavTabs from './NavTabs';

export const TabContext = React.createContext();

export const getTabId = tabLabel => tabLabel.toLowerCase();

const Tabs = ({ tabLabels, tabIds, onTabChange, activeTab, defaultTab, children }) => (
  <div className="nav-tabs-custom">
    <NavTabs
      tabLabels={tabLabels}
      tabIds={tabIds}
      onTabChange={onTabChange}
      activeTab={activeTab || defaultTab}
    />
    <TabContext.Provider value={activeTab || defaultTab}>
      <div className="tab-content">{children}</div>
    </TabContext.Provider>
  </div>
);

Tabs.propTypes = {
  /**
   * Labels of tabs
   */
  tabLabels: PropTypes.arrayOf(PropTypes.string),
  /**
   * Ids of tabs --> must be the same length with labels
   * If not provided --> use getTabId()
   */
  tabIds: PropTypes.arrayOf(PropTypes.string),
  /**
   * Callback that receives the target tabId
   */
  onTabChange: PropTypes.func,
  /**
   * Id of the active tab
   */
  activeTab: PropTypes.string,
  /**
   * Id of the tab to open when activeTab does not exist
   */
  defaultTab: PropTypes.string,
};

export default Tabs;
