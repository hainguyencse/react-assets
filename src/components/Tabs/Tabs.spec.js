import React from 'react';
import Tabs from './index';
import TabContent from './TabContent';

describe('Tabs', () => {
  test('should render properly', () => {
    // eslint-disable-next-line
    const wrapper = shallow(<Tabs tabLabels={['Tab 1', 'Tab 2']} />);

    // eslint-disable-next-line
    expect(wrapper).toMatchSnapshot();
  });

  test('should render right number of tabs', () => {
    // eslint-disable-next-line
    const wrapper = shallow(<Tabs tabLabels={['Tab 1', 'Tab 2']} />);

    expect(
      wrapper
        .find('NavTabs')
        .dive()
        .find('Tab')
    ).toHaveLength(2);
  });

  test('should render active tab properly', () => {
    // eslint-disable-next-line
    const wrapper = mount(
      <Tabs tabLabels={['Tab 1', 'Tab 2']} activeTab="tab 2">
        <TabContent tabId="tab 1" component={() => <div />} />
        <TabContent tabId="tab 2" component={() => <div />} />
      </Tabs>
    );

    expect(
      wrapper
        .find('Tab')
        .find({ id: 'tab 2' })
        .find('li')
        .hasClass('active')
    ).toEqual(true);

    expect(
      wrapper
        .find('TabContent')
        .find({ tabId: 'tab 2' })
        .find('.tab-pane')
        .hasClass('active')
    ).toEqual(true);

    wrapper.unmount();
  });
});
