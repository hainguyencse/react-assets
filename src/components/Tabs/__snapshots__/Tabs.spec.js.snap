// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tabs
    tabLabels={
      Array [
        "Tab 1",
        "Tab 2",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <NavTabs
          tabLabels={
            Array [
              "Tab 1",
              "Tab 2",
            ]
          }
        />,
        <Context.Provider>
          <div
            className="tab-content"
          />
        </Context.Provider>,
      ],
      "className": "nav-tabs-custom",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "activeTab": undefined,
          "onTabChange": undefined,
          "tabIds": undefined,
          "tabLabels": Array [
            "Tab 1",
            "Tab 2",
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div
            className="tab-content"
          />,
          "value": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "tab-content",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": Object {
          "$$typeof": Symbol(react.provider),
          "_context": Object {
            "$$typeof": Symbol(react.context),
            "Consumer": [Circular],
            "Provider": [Circular],
            "_calculateChangedBits": null,
            "_changedBits": 0,
            "_changedBits2": 0,
            "_currentRenderer": null,
            "_currentRenderer2": null,
            "_currentValue": undefined,
            "_currentValue2": undefined,
            "_defaultValue": undefined,
          },
        },
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <NavTabs
            tabLabels={
              Array [
                "Tab 1",
                "Tab 2",
              ]
            }
          />,
          <Context.Provider>
            <div
              className="tab-content"
            />
          </Context.Provider>,
        ],
        "className": "nav-tabs-custom",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeTab": undefined,
            "onTabChange": undefined,
            "tabIds": undefined,
            "tabLabels": Array [
              "Tab 1",
              "Tab 2",
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div
              className="tab-content"
            />,
            "value": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "tab-content",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": Object {
            "$$typeof": Symbol(react.provider),
            "_context": Object {
              "$$typeof": Symbol(react.context),
              "Consumer": [Circular],
              "Provider": [Circular],
              "_calculateChangedBits": null,
              "_changedBits": 0,
              "_changedBits2": 0,
              "_currentRenderer": null,
              "_currentRenderer2": null,
              "_currentValue": undefined,
              "_currentValue2": undefined,
              "_defaultValue": undefined,
            },
          },
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
